File: app/src/main/java/com/kaenozu/aimitsumori/domain/normalization/Normalizer.kt
package com.kaenozu.aimitsumori.domain.normalization

import com.kaenozu.aimitsumori.data.local.CategoryMaster
import com.kaenozu.aimitsumori.domain.model.InclusionStatus
import com.kaenozu.aimitsumori.domain.model.NormalizedLine
import com.kaenozu.aimitsumori.domain.model.NormalizedQuote
import com.kaenozu.aimitsumori.domain.model.Project
import com.kaenozu.aimitsumori.domain.model.QuoteLineItem

class Normalizer {
    fun normalize(project: Project): List<NormalizedQuote> =
        project.quotes.map { quote ->
            val itemsByCategory = quote.lineItems.groupBy(QuoteLineItem::categoryId)
            NormalizedQuote(
                quoteId = quote.id,
                contractorName = quote.contractorName,
                totalAmountYen = quote.totalAmountYen,
                lines = CategoryMaster.categories.map { category ->
                    normalizeCategory(
                        categoryId = category.id,
                        items = itemsByCategory[category.id].orEmpty(),
                    )
                },
            )
        }

    private fun normalizeCategory(
        categoryId: String,
        items: List<QuoteLineItem>,
    ): NormalizedLine {
        val category = CategoryMaster.require(categoryId)
        if (items.isEmpty()) {
            return NormalizedLine(
                category = category,
                inclusionStatus = InclusionStatus.UNKNOWN,
                amountYen = null,
                quantity = null,
                unit = null,
                specification = null,
                sourceLineItemIds = emptyList(),
                uncertaintyReasons = listOf("見積明細に記載がありません"),
            )
        }

        val reasons = mutableListOf<String>()
        val distinctStatuses = items.map(QuoteLineItem::inclusionStatus).distinct()
        val status = if (distinctStatuses.size == 1) {
            distinctStatuses.single()
        } else {
            reasons += "同じカテゴリ内で含有状態が一致していません"
            InclusionStatus.UNKNOWN
        }

        val amount = when {
            items.all { it.amountYen == null } -> null
            items.any { it.amountYen == null } -> {
                reasons += "金額未記載の明細を含みます"
                items.mapNotNull(QuoteLineItem::amountYen).sum()
            }
            else -> items.sumOf { requireNotNull(it.amountYen) }
        }

        val quantityValues = items.mapNotNull(QuoteLineItem::quantity).distinct()
        val unitValues = items.mapNotNull { it.unit?.trim()?.takeIf(String::isNotEmpty) }.distinct()
        val quantity = quantityValues.singleOrNull()
        val unit = unitValues.singleOrNull()
        if (category.quantityExpected && (quantity == null || unit == null)) {
            reasons += "数量または単位が不明です"
        }
        if (quantityValues.size > 1 || unitValues.size > 1) {
            reasons += "複数明細の数量・単位を単一値へ統合できません"
        }

        val specificationValues = items
            .mapNotNull { it.specification?.trim()?.takeIf(String::isNotEmpty) }
            .distinct()
        val specification = specificationValues.singleOrNull()
        if (category.specificationExpected && specification == null) {
            reasons += "仕様・型番が不明です"
        }
        if (specificationValues.size > 1) {
            reasons += "複数の仕様・型番が混在しています"
        }
        if (status == InclusionStatus.UNKNOWN) {
            reasons += "見積に含むかどうか不明です"
        }

        return NormalizedLine(
            category = category,
            inclusionStatus = status,
            amountYen = amount,
            quantity = quantity,
            unit = unit,
            specification = specification,
            sourceLineItemIds = items.map(QuoteLineItem::id),
            uncertaintyReasons = reasons.distinct(),
        )
    }
}

File: lib/normalizer.dart
/// ファイルパス: lib/normalizer.dart
/// 見積明細を共通カテゴリへ正規化するロジック。
library;

import 'data/category_master.dart';
import 'models.dart';
import 'unit_normalizer.dart';

class Normalizer {
  List<NormalizedQuote> normalize(Project project) {
    return project.quotes.map((quote) {
      final itemsByCategory = <String, List<QuoteLineItem>>{};
      for (final item in quote.lineItems) {
        itemsByCategory.putIfAbsent(item.categoryId, () => []).add(item);
      }
      return NormalizedQuote(
        quoteId: quote.id,
        contractorName: quote.contractorName,
        totalAmountYen: quote.totalAmountYen,
        lines: CategoryMaster.categories.map((category) {
          return _normalizeCategory(
            category: category,
            items: itemsByCategory[category.id] ?? const [],
          );
        }).toList(growable: false),
      );
    }).toList(growable: false);
  }

  NormalizedLine _normalizeCategory({
    required CategoryDefinition category,
    required List<QuoteLineItem> items,
  }) {
    if (items.isEmpty) {
      return NormalizedLine(
        category: category,
        inclusionStatus: InclusionStatus.unknown,
        uncertaintyReasons: const ['見積明細に記載がありません'],
      );
    }

    final reasons = <String>[];
    final distinctStatuses = items
        .map((item) => item.inclusionStatus)
        .toSet();
    final status = distinctStatuses.length == 1
        ? distinctStatuses.single
        : InclusionStatus.unknown;
    if (distinctStatuses.length > 1) {
      reasons.add('同じカテゴリ内で含有状態が一致していません');
    }

    final int? amount;
    if (items.every((item) => item.amountYen == null)) {
      amount = null;
    } else {
      if (items.any((item) => item.amountYen == null)) {
        reasons.add('金額未記載の明細を含むため、表示額は記載分のみです');
      }
      amount = items
          .where((item) => item.amountYen != null)
          .fold<int>(0, (sum, item) => sum + item.amountYen!);
    }

    final quantityItems = items.where((item) => item.quantity != null).toList();
    final normalizedUnits = items
        .map((item) => UnitNormalizer.normalize(item.unit))
        .whereType<String>()
        .toSet();

    double? quantity;
    String? unit;
    if (quantityItems.length != items.length) {
      if (quantityItems.isNotEmpty) {
        reasons.add('数量未記載の明細を含むため、数量を合算できません');
      }
    } else if (normalizedUnits.length == 1) {
      unit = normalizedUnits.single;
      quantity = quantityItems.fold<double>(
        0,
        (sum, item) => sum + item.quantity!,
      );
    } else if (normalizedUnits.isEmpty && items.length == 1) {
      quantity = items.single.quantity;
    } else {
      reasons.add('複数明細の単位が一致しないため、数量を合算できません');
    }

    if (category.quantityExpected && (quantity == null || unit == null)) {
      reasons.add('数量または単位が不明です');
    }

    final specificationValues = items
        .where((item) => item.specification?.trim().isNotEmpty == true)
        .map((item) => item.specification!.trim())
        .toSet();
    final specification = specificationValues.length == 1
        ? specificationValues.single
        : null;
    if (category.specificationExpected && specification == null) {
      reasons.add('仕様・型番が不明です');
    }
    if (specificationValues.length > 1) {
      reasons.add('複数の仕様・型番が混在しています');
    }
    if (status == InclusionStatus.unknown) {
      reasons.add('見積に含むかどうか不明です');
    }

    return NormalizedLine(
      category: category,
      inclusionStatus: status,
      amountYen: amount,
      quantity: quantity,
      unit: unit,
      specification: specification,
      sourceLineItemIds: items.map((item) => item.id).toList(growable: false),
      uncertaintyReasons: reasons.toSet().toList(growable: false),
    );
  }
}
